### PR TITLE
Stencil buffer access + misc

### PR DESCRIPTION
**Pull request by guycalledfrank**
Stencil buffer access. Context is always created with the "stencil" option now. New device functions:
- supportsStencil property tells if stencil buffer is supported;
- setStencilTest(bool) enables/disables stencil;
- setStencilFunc(func, ref, mask) configures the stencil test;
- setStencilFuncFront
- setStencilFuncBack
- setStencilOperation(fail, zfail, zpass) configures how stencil buffer should be modified.
- setStencilOperationFront
- setStencilOperationBack

New class:
- StencilParameters: contains info for stencilfunc/stencilop.

New material properties:
- stencilFront;
- stencilBack.

You set StencilParameters objects to these.

New camera component property:
- clearStencilBuffer (bool, true by default)

Misc:
- device.supportsMsaa tells if MSAA is supported/enabled;
- **[BREAKING]** screen depth is changed, it's now linear view depth. Makes more sense/easier to reconstruct position. Some code using depth may work differently. Particles (softening) are patched.
- drawQuadWithShader can be used with custom blending.

### CODE REVIEW

File: src/scene/material.js [hunk 1]
```diff
@@ -202,6 +206,8 @@ pc.extend(pc, function () {
 
         clone.depthTest = this.depthTest;
         clone.depthWrite = this.depthWrite;
+        clone.stencilFront = this.stencilFront;
```
Comment (willeastcott) [inline src/scene/material.js]: Shouldn't you be copy this rather than referencing?

File: src/scene/material.js [hunk 2]
```diff
@@ -400,7 +406,18 @@ pc.extend(pc, function () {
         this.shader = shader;
     };
 
+    var StencilParameters = function (options) {
```
Comment (willeastcott) [inline src/scene/material.js]: Needs documenting.

File: src/scene/material.js (hunk 1 above)
Comment (guycalledfrank) [inline src/scene/material.js]: Hmm yeah. I'll add copy/clone methods for StencilParameters.

File: src/graphics/graphics.js [hunk 3]
```diff
@@ -666,7 +666,25 @@
         UNIFORMTYPE_TEXTURECUBE: 16,
         UNIFORMTYPE_FLOATARRAY: 17,
 
-        SHADERTAG_MATERIAL: 1
+        SHADERTAG_MATERIAL: 1,
+
+        FUNC_NEVER: 0,
```
Comment (willeastcott) [inline src/graphics/graphics.js]: I ordered the 'enums' in this file alphabetically, at least based on the word before the first underscore. Not a big deal but....

Comment (guycalledfrank) [inline src/graphics/graphics.js]: Heh, never noticed that. Can reorder.

File: src/scene/material.js (hunk 1 above)
Comment (guycalledfrank) [inline src/scene/material.js]: Added clone

File: src/scene/material.js (hunk 2 above)
Comment (guycalledfrank) [inline src/scene/material.js]: Documented

File: src/graphics/graphics.js (hunk 3 above)
Comment (guycalledfrank) [inline src/graphics/graphics.js]: Reordered

**Comment by vkalpias** [comment]
Do you know if this breaks any post effects that use the depth buffer?

**Comment by guycalledfrank** [comment]
Can't find any depth-based post-effect on the Store.
Things like DoF should work, because they work only with the screen depth. Problems can arise when you compare fragment depth with screen depth (as in soft particles), because then you need to provide your fragment depth in the same format.

**Pull request merged**